qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+    for (word_counts_t::const_iterator it_query = query_counts.begin();
+        it_query != query_counts.end(); ++it_query)
+    {
+        // compute the idf
+        double idf = get_idf(it_query->first);
+
+        // query vector
+        query_vector.push_back(
+            (0.5 + 0.5 * it_query->second / max_query) * idf);
+
+        // doc vector - if word is in doc then compute it, otherwise it's 0
+        word_counts_t::const_iterator got = doc_counts.find(it_query->first);
+        if (got != doc_counts.end())
+            doc_vector.push_back(idf * got->second);
+        else
+            doc_vector.push_back(0.0);
+    }
+
+    // need to normalize the document vector with the euclidian length
+    // the doc_vector variable only includes elements in the doc vector
+    // corresponding to query words but we need the length of the full
+    // vector here.  therefore need to iterate through doc_counts
+    double doc_vector_len = 0.0;
+    for (word_counts_t::const_iterator it_doc = doc_counts.begin();
+        it_doc != doc_counts.end(); ++it_doc)
+    {
+        double p = ((double) it_doc->second) * get_idf(it_doc->first);
+        doc_vector_len += p * p;
+    }
+    doc_vector_len = sqrt(doc_vector_len);
+
+    // finally compute the score
+    double score = 0.0;
+    for (std::size_t k = 0; k < doc_vector.size(); ++k)
+        score += query_vector[k] * doc_vector[k];
+    return score / doc_vector_len;
+}
+
+
+// magic constants for okapi_bm25 function.  we can allow these to
+// be configurable down the road if needed.
+#define BM25_K1 1.6
+#define BM25_B 0.75
+
+double QDR::okapi_bm25(
+    const word_counts_t& doc_counts, const word_counts_t& query_counts,
+    const double& Ld)
+{
+    /// Okapi BM25 ranking function
+    // See "An Introduction to Information Retrieval" by Manning,
+    //    Raghavan, Schutz.  Section 11.4.3 (page 233), eqn 11.32
+    // the ranking function is:
+    // SUM_{t in query} log(N / df[t]) * (k1 + 1) * tf[td] /
+    //                  (k1 * ((1 - b) + b * (Ld / Lave)) + tf[td])
+    //
+    // where N = number docs in corpus
+    //  df[t]  =  number docs with term t
+    //  tf[td] = number of occurrences of term t in this document
+    //  Ld = length of this document (# words)
+    //  Lave = average length of documents in corpus
+    //  k1, b = free parameters, empirically set to about
+    //  k1 = 1.2 - 2.0
+    //  b = 0.75
+
+    double Lave = ((double) nwords) / ((double) total_docs);
+    double Ld_Lave = Ld / Lave;
+
+    double score = 0.0;
+    for (word_counts_t::const_iterator it_query = query_counts.begin();
+        it_query != query_counts.end(); ++it_query)
+    {
+        word_counts_t::const_iterator got = doc_counts.find(it_query->first);
+        if (got != doc_counts.end())
+        {
+            double idf = get_idf(it_query->first);
+            double tf_doc = (double) got->second;
+            score += ((double) it_query->second) * idf * (BM25_K1 + 1.0) *
```

Comment:
This one is a little less clear, but I'd still probably drop the cast. `idf` and `BM25_K1 + 1.0` are both double, so the resulting expression must also be double.